pyomo/common/unittest.py

```diff
@@ -547,3 +554,404 @@ def assertRaisesRegex(self, expected_exception, expected_regex, *args, **kwargs)
+        super().__init__(test)
+
+    def initialize_dependencies(self):
+        # Note: as a rule, pyomo.common is not allowed to import from
+        # the rest of Pyomo.  we permit it here because a) this is not
+        # at module scope, and b) there is really no better / more
+        # logical place in pyomo to put this code.
+        from pyomo.opt import check_available_solvers
+
+        cls = self.__class__
+        #
+        # Initialize the availability data
+        #
+        solvers_used = set(sum(list(cls.solver_dependencies.values()), []))
+        available_solvers = check_available_solvers(*solvers_used)
+        cls.solver_available = {
+            solver_: (solver_ in available_solvers) for solver_ in solvers_used
+        }
+
+        cls.package_available = {}
+        cls.package_modules = {}
+        packages_used = set(sum(list(cls.package_dependencies.values()), []))
+        for package_ in packages_used:
+            pack, pack_avail = attempt_import(package_, defer_check=False)
+            cls.package_available[package_] = pack_avail
+            cls.package_modules[package_] = pack
+
+    @classmethod
+    def _find_tests(cls, test_dirs, pattern):
+        test_tuples = []
+        for testdir in test_dirs:
+            # Find all pattern files in the test directory and any immediate
+            # sub-directories
+            for fname in list(glob.glob(os.path.join(testdir, pattern))) + list(
+                glob.glob(os.path.join(testdir, '*', pattern))
+            ):
+                test_file = os.path.abspath(fname)
+                bname = os.path.basename(test_file)
+                dir_ = os.path.dirname(test_file)
+                name = os.path.splitext(bname)[0]
+                tname = os.path.basename(dir_) + '_' + name
+
+                suffix = None
+                # Look for txt and yml file names matching py file names. Add
+                # a test for any found
+                for suffix_ in ['.txt', '.yml']:
+                    if os.path.exists(os.path.join(dir_, name + suffix_)):
+                        suffix = suffix_
+                        break
+                if suffix is not None:
+                    tname = tname.replace('-', '_')
+                    tname = tname.replace('.', '_')
+
+                    # Create list of tuples with (test_name, test_file, baseline_file)
+                    test_tuples.append(
+                        (tname, test_file, os.path.join(dir_, name + suffix))
+                    )
+
+        # Ensure a deterministic test ordering
+        test_tuples.sort()
+        return test_tuples
+
+    @classmethod
+    def gather_tests(cls, test_dirs):
+        # Find all .sh files in the test directories
+        sh_test_tuples = cls._find_tests(test_dirs, '*.sh')
+
+        # Find all .py files in the test directories
+        py_test_tuples = cls._find_tests(test_dirs, '*.py')
+
+        # If there is both a .py and a .sh, defer to the sh
```

Comment:
Historical reasoning for this?

Reply:
The short answer is that in the context of testing Pyomo, you could envision driving python through a shell script, but not a shell script through python.

The actual case that motivated this is an example that intentionally raises an exception.  The shell script tester allows us to test that the exception got raised (because it is part of the stderr) ... and because the shell script doesn't check the python return code.  BUT, if we imported/tested the python directly, that test would fail (due to the exception)... so this policy allows us to "mask" some python files by adding a shell script driver.